share/pvss.go

```diff
@@ -0,0 +1,174 @@
+
+	// Create secret sharing polynomial
+	priPoly := NewPriPoly(pv.suite, pv.t, secret, random.Stream)
+
+	// Create secret set of shares
+	priShares := priPoly.Shares(n)
+
+	// Create public polynomial commitments with respect to basis H
+	pubPoly := priPoly.Commit(pv.h)
+
+	// Prepare data for encryption consistency proofs ...
+	indices := make([]int, n)
+	values := make([]abstract.Scalar, n)
+	H := make([]abstract.Point, n)
+	for i := 0; i < n; i++ {
+		indices[i] = priShares[i].I
+		values[i] = priShares[i].V
+		H[i] = pv.h
+	}
+
+	// Create NIZK discrete-logarithm equality proofs
+	proofs, _, sX, err := proof.NewDLEQProofBatch(pv.suite, H, X, values)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for i := 0; i < n; i++ {
+		ps := &PubShare{indices[i], sX[i]}
+		encShares[i] = &PubVerShare{*ps, *proofs[i]}
+	}
+
+	return encShares, pubPoly, nil
+}
+
+func (pv *PVSS) verifyEncShares(H abstract.Point, X []abstract.Point, polys []*PubPoly, encShares []*PubVerShare) ([]abstract.Point, []*PubVerShare, error) {
+
+	if len(X) != len(polys) && len(polys) != len(encShares) {
+		return nil, nil, errorDifferentLengths
+	}
+
+	// Recover commits from polynomials
+	n := len(X)
+	sH := make([]*PubShare, n)
+	for i := 0; i < n; i++ {
+		sH[i] = polys[i].Eval(encShares[i].S.I)
+	}
+
+	var goodKeys []abstract.Point
+	var goodShares []*PubVerShare
+	for i := 0; i < n; i++ {
+		if encShares[i].P.Verify(pv.suite, H, X[i], sH[i].V, encShares[i].S.V) {
+			goodKeys = append(goodKeys, X[i])
+			goodShares = append(goodShares, encShares[i])
+		}
+	}
+
+	return goodKeys, goodShares, nil
+}
+
+// DecShares first verifies the given encrypted shares against their encryption
+// consistency proofs, i.e., it checks that every share sX satisfies log_H(sH)
+// == log_X(sX). Afterwards all valid shares are decrypted and decryption consistency
+// proofs are created.
+func (pv *PVSS) DecShares(H abstract.Point, X []abstract.Point, polys []*PubPoly, x abstract.Scalar, encShares []*PubVerShare) ([]abstract.Point, []*PubVerShare, []*PubVerShare, error) {
+
+	if len(X) != len(polys) && len(polys) != len(encShares) {
+		return nil, nil, nil, errorDifferentLengths
+	}
+
+	goodKeys, goodEncShares, err := pv.verifyEncShares(H, X, polys, encShares)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var decShares []*PubVerShare
+	for _, s := range goodEncShares {
+
+		G := pv.suite.Point().Base()
+		V := pv.suite.Point().Mul(s.S.V, pv.suite.Scalar().Inv(x)) // decryption: x^{-1} * (xS)
+
+		P, _, _, err := proof.NewDLEQProof(pv.suite, G, V, x)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		ps := &PubShare{s.S.I, V}
+		decShares = append(decShares, &PubVerShare{*ps, *P})
+	}
+
+	return goodKeys, goodEncShares, decShares, nil
+}
+
+func (pv *PVSS) verifyDecShares(G abstract.Point, X []abstract.Point, encShares []*PubVerShare, decShares []*PubVerShare) ([]*PubVerShare, error) {
+
+	if len(X) != len(encShares) || len(encShares) != len(decShares) {
```

Comment:
question: now it's the contrary - only one inequality is enough to trigger the error - is this correct here?

doc: please comment why once you need `||` and twice you need `&&`